File: 004_Median_of_Two_Sorted_Arrays/solution.js
/*
 * There are two sorted arrays nums1 and nums2 of size m and n respectively.

 Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).

 Example 1:
 nums1 = [1, 3]
 nums2 = [2]
 The median is 2.0.

 Example 2:
 nums1 = [1, 2]
 nums2 = [3, 4]
 The median is (2 + 3)/2 = 2.5
*/

module.exports = {
    // Relay on the embedded sort algorithms
    // Lazy and cunning
    findMedianSortedArrays1: function(nums1, nums2){

        let nums = nums1.concat(nums2);
        nums.sort(function(a,b){return a - b});
        console.log(nums);
        let len  = nums.length;
        if(len % 2 === 0){
            let n  = len / 2;
            return (nums[n-1] + nums[n])/2;
        }else{
            return nums[Math.floor(len / 2)];
        }
    },
    // dived into two parts, with same length
    findMedianSortedArrays2: function(nums1, nums2){
        //swap
        if(nums1.length > nums2.length){
            let tmp = nums2;
            nums2 = nums1;
            nums1 = tmp;
        }

        let m = nums1.length;
        let n = nums2.length;
        let iLeft  = 0;
        let iRight =  m-1;
        while(true){
            let i = Math.floor((iLeft + iRight)/2); // initial
            let j = Math.floor((m+n-i*2)/2)-1;
            console.log("i="+i+";j="+j);
            if(i === -1){
                return [nums2[j-1], nums2[j]];
            }else if( nums1[i] <= nums2[j+1] &&
                     (i === m - 1 || nums2[j] <= nums1[i+1])){//TODO Boundary i = m!!
                return [nums1[i],nums2[j]];
            }else if(nums1[i] > nums2[j+1]){//  Move to left
                iRight = i - 1;
            }else if(nums2[j] > nums1[i+1]){ // Move to Right
                iLeft = i + 1;
            }
        }
    }
}

File: 004_Median_of_Two_Sorted_Arrays/Java/Solution.java
class Solution {
    // Runtime: 4 ms, faster than 38.59% of Java online submissions for Median of Two Sorted Arrays.
    // Memory Usage: 46 MB, less than 38.63% of Java online submissions for Median of Two Sorted Arrays.
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        // Swap nums to make sure nums1.length < nums2.length
        if(nums1.length > nums2.length){
            int[] temp = nums1;
            nums1 = nums2;
            nums2 = temp;
        }

        double result = 0;
        int i = 0, j = 0;
        int m, n;
        boolean isEven;

        m = nums1.length;
        n = nums2.length;

        isEven = (m+n)%2 == 0;

        int i_left,i_right, j_left, j_right;
        int max_left, min_right;
        for(;i<m+1;i++){
            //Boundary of i = m, j = n!
            j = (m+n+1)/2 - i;
            if(m == 0){ // First Array is empty
                if(j == 0){ // Second Array is empty
                    result = 0;
                }else if(j == n){ // Second Array contains only one element
                    result  = nums2[j-1];
                }else{
                    result = isEven ? (nums2[j-1] + nums2[j])/2.0 : nums2[j-1] ;
                }
                break;
            }
            // Compare digits
            if(i == 0){
                max_left = nums2[j-1];
            }else if( j == 0){
                max_left = nums1[i-1];
            }else{
                max_left = Math.max(nums1[i-1], nums2[j-1]);
            }

            if(i == m){
                min_right = nums2[j];
            }else if( j == n){
                min_right = nums1[i];
            }else{
                min_right = Math.min(nums1[i], nums2[j]);
            }

            if(max_left <= min_right){ // find the position
                if(isEven){
                    result = (max_left + min_right) / 2.0;
                }else{
                    result = max_left;
                }
            }
        }
        return result;
    }

    public static void main(String[] args){
        Solution sol = new Solution();
        int[] n1 = new int[]{1,2};
        int[] n2 = new int[]{3,4};
        System.out.println(sol.findMedianSortedArrays(n1,n2));
    }
}
